Automatic FX-mixer strip management #1215

spechtstatt opened the pull request:
A very first draft of an automatic mixer strip management as proposed by @unfa  which allows an automatic linking of tracks to fx mixer strips. 

Works so far for the song editor - bbeditor not tested yet and also the automatic sorting is not implemented yet.

![AutoFXFeature_Draft01](https://user-images.githubusercontent.com/93736385/168491846-1f660026-e231-472a-9243-4f2328ed2750.gif)

spechtstatt commented:
Would be great to get some feedback about the functionality in general and also regarding my implementation approach. I tried to move some functionality to the FxMixer and use the FxMixerView as a kind of entry point for the track operations. 

If this makes sense I can work on resolving the conflicts.

spechtstatt commented:
### Open Tasks

- [ ] rebase 
- [ ] make buildable again
- [ ] test bbeditor 
- [ ] automatic sorting 
- [ ] add global config switch
- [ ] button for enabling/disabling the functionilty in the mixer

spechtstatt commented:
It would also be possible in the course of this PR to provide some additional information directly inside the mixer as it is proposed by #1214 because this PR is collecting this information anyway.

Like e.g. count of connections to the mixer channel or information about the area (pattern vs. song). Auto linked channels are currently identified by changing the text color to green. 

Maybe someone has some ideas for presenting all this information inside the UI.

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -185,11 +272,61 @@ int MixerView::addNewChannel()
+	mix->processAssignedTracks([&, list](Track * track, IntModel * model, MixerChannel * channel)
+	mutable {
+		(void) track;		
+		if (channel == NULL) return;		
```
```suggestion
		if (channel == nullptr) return;		
```

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -419,79 +547,86 @@ void MixerView::deleteChannel(int index)
+	MixerChannel *  channel = mix->mixerChannel(index);
+	if (!channel->m_autoTrackLinkModel.value()) return;
+
+	Track * trackFound = NULL;
```
```suggestion
	Track * trackFound = nullptr;
```

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -419,79 +547,86 @@ void MixerView::deleteChannel(int index)
 
-	//Check all channels except master, delete those with no incoming sends
-	for(int i = m_mixerChannelViews.size()-1; i > 0; --i)
+	if (trackFound != NULL)
```
```suggestion
	if (trackFound != nullptr)
```

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -170,6 +170,93 @@ MixerView::~MixerView()
+{
+	Mixer * mix = Engine::mixer();
+	IntModel * model = mix->getChannelModelByTrack(track);
+	if (model != NULL)
```
```suggestion
	if (model != nullptr)
```

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -170,6 +170,93 @@ MixerView::~MixerView()
+{
+	Mixer * mix = Engine::mixer();
+	IntModel * model = mix->getChannelModelByTrack(track);
+	if (model != NULL)
```
```suggestion
	if (model != nullptr)
```

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -170,6 +170,93 @@ MixerView::~MixerView()
+{
+	Mixer * mix = Engine::mixer();
+	IntModel * model = Engine::mixer()->getChannelModelByTrack(track);
+	if (model != NULL)
```
```suggestion
	if (model != nullptr)
```

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -170,6 +170,93 @@ MixerView::~MixerView()
+{
+	Mixer * mix = Engine::mixer();
+	IntModel * model = mix->getChannelModelByTrack(track);
+	if ( model != NULL)
```
```suggestion
	if ( model != nullptr)
```

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -185,11 +272,61 @@ int MixerView::addNewChannel()
+{
+
+	Mixer * mix = Engine::mixer();
+	QList<int> *list = new QList<int>();
```
Should this really be allocated on the heap? A simple ``std::vector`` on the stack would work fine here, no?

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -185,11 +272,61 @@ int MixerView::addNewChannel()
+
+	// bubblesort here because the list is normally almost ordered
+	int n = list->length();
+	bool swapped = false;
```
``swapped`` is not reassigned any value in this do-while loop. Wouldn't this just run once? Unless the intent is something different, I think this can be simplified.

spechtstatt replied:
The sorting approach is not working out as I thought - therefore the return - but I had no time yet to think again about it.

sakertooth commented on src/core/Mixer.cpp:
```diff
@@ -368,74 +348,122 @@ void Mixer::deleteChannel( int index )
+	std::vector<int> used(m_mixerChannels.size(), 0);
+	processAssignedTracks([&used](Track * track, IntModel * model, MixerChannel * channel)
+	mutable {
+		(void) track;
```
Is this necessary? I see you've done something like ``(void) <parameter>`` in other places as well, even when the parameter in question isn't used by the lambda.

spechtstatt replied:
I just wanted to avoid the warning and signal that the parameter is not used by intent but it is not necessary.

sakertooth commented on src/core/Mixer.cpp:
```diff
@@ -368,74 +348,122 @@ void Mixer::deleteChannel( int index )
+{
+	if (mixerChannel( index )->m_autoTrackLinkModel.value()) return true;
 
+	std::vector<int> usedChannelCounts = getUsedChannelCounts();
```
```suggestion
```

sakertooth commented on src/core/Mixer.cpp:
```diff
@@ -368,74 +348,122 @@ void Mixer::deleteChannel( int index )
+	if (mixerChannel( index )->m_autoTrackLinkModel.value()) return true;
 
+	std::vector<int> usedChannelCounts = getUsedChannelCounts();
+	return usedChannelCounts[index] == 1;
```
```suggestion
	return getUsedChannelCounts()[index] == 1;
```

sakertooth replied:
I find this to be more compact. Up to you if you decide to make the change.

sakertooth commented on src/core/Mixer.cpp:
```diff
@@ -368,74 +348,122 @@ void Mixer::deleteChannel( int index )
+
+IntModel * Mixer::getChannelModelByTrack(Track * track)
+{
+	if( track->type() == Track::InstrumentTrack )
```
You can also use a switch-case.

spechtstatt replied:
Yes - but normally I would not use a switch statement to differ between only two outcomes.

sakertooth commented on src/core/Mixer.cpp:
```diff
@@ -285,46 +286,25 @@ void Mixer::deleteChannel( int index )
+	processAssignedTracks( [index](Track * track, IntModel * model, MixerChannel * channel)
 	{
-		if( t->type() == Track::InstrumentTrack )
+		(void) track;
```
Why is this here? I'm not too sure of its purpose.

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -419,79 +547,86 @@ void MixerView::deleteChannel(int index)
-		//setting this to 0 is a safe default (for tracks that don't sent to the mixer).
-		int channel = 0;
-		if (t->type() == Track::InstrumentTrack)
+		if ((inUse[i]==0) && Engine::mixer()->mixerChannel(i)->m_receives.isEmpty())
```
```suggestion
		if ((inUse[i]==0) && mix->mixerChannel(i)->m_receives.isEmpty())
```

sakertooth commented on src/core/Mixer.cpp:
```diff
@@ -285,46 +286,25 @@ void Mixer::deleteChannel( int index )
 	{
-		if( t->type() == Track::InstrumentTrack )
+		(void) track;
+		(void) channel;
```
Same with this one.

sakertooth commented on src/gui/MixerView.cpp:
```diff
@@ -170,6 +170,93 @@ MixerView::~MixerView()
+		std::vector<bool> used(m_mixerChannelViews.size(), false);
+		bool needUpdate = false;
+		std::vector<int> usedChannelCounts = mix->getUsedChannelCounts();
+		for(unsigned long i = 0; i< usedChannelCounts.size();i++)
```
```suggestion
		for(unsigned long i = 0; i < usedChannelCounts.size(); i++)
```

sakertooth commented on src/gui/widgets/MixerLine.cpp:
```diff
@@ -244,14 +247,31 @@ void MixerLine::mouseDoubleClickEvent( QMouseEvent * )
 {
-	QPointer<CaptionMenu> contextMenu = new CaptionMenu( Engine::mixer()->mixerChannel( m_channelIndex )->m_name, this );
+	Mixer * mix =  Engine::mixer();
+	QPointer<CaptionMenu> contextMenu = new CaptionMenu( mix->mixerChannel( m_channelIndex )->m_name, this );
```
Why not just:
```suggestion
	CaptionMenu* contextMenu = new CaptionMenu( mix->mixerChannel( m_channelIndex )->m_name, this );
```

sakertooth commented on src/gui/widgets/MixerLine.cpp:
```diff
@@ -262,21 +282,44 @@ void MixerLine::contextMenuEvent( QContextMenuEvent * )
+	if (!autoTrackLink)
+	{
+		colorMenu.setIcon(embed::getIconPixmap("colorize"));
+		colorMenu.addAction(tr("Change"), this, SLOT(selectColor()));
```
I find ``SLOT()`` to be redundant mainly because any function can act as a slot now. If you want to be consistent with the current codebase at the moment, then I would say it is not that big of an issue.

https://doc.qt.io/qt-5/signalsandslots.html#:~:text=You%20can%20also%20connect%20to%20functors%20or%20C%2B%2B11%20lambdas%3A

spechtstatt replied:
Yes I know now but I just thought it is more consistent with what is already there.

sakertooth reviewed (changes requested):
Just a few changes and a quick review from me. I had some questions as to what certain lines were doing, especially the ``(void) track`` and ``(void) channel`` ones.

Veratil commented on src/gui/MixerView.cpp:
```diff
@@ -185,11 +272,61 @@ int MixerView::addNewChannel()
+	return;
+
+
+	// bubblesort here because the list is normally almost ordered
```
why not `std::sort` ?

spechtstatt replied:
Because I am C++ beginner :-)

I will look into it.

sakertooth replied:
I was thinking of ``std::sort`` as well, but I was not sure how the default swap behavior would be changed to use ``swapChannels`` instead.

spechtstatt replied:
Because it is described as a non stable sorting algorithm (https://en.cppreference.com/w/cpp/algorithm/sort) I could imagine that this is maybe a quicksort.

When I was thinking about the requirements I considered the following: the items are already nearly sorted, there are relatively few items and the sorting algorithm should swap in a mostly linear fashion. 

There is a good animation for comparison here:
https://www.toptal.com/developers/sorting-algorithms/nearly-sorted-initial-order

Of course it could be that there is in fact no difference if only one single element is out of order as it would be in this case.

Monospace-V commented:
I noticed a few things, could you clarify these?
Setting an fx mixer colour before linking makes the FX mixer's colour permanent unless the track's colour is changed (not reset, changed). This is a **bug**.
Could we get a direct "Assign and link to new Mixer channel" option, so we don't have to do it twice? It seems useful. Also an "Link to Mixer channel" in the context menu of track itself?
After linking, colours are linked but can be edited only from the track's context menu. However, some of us color-code multiple channels consecutively in the mixer itself based on the instrument type, and it could get annoying having to search. Can this be made two-way, being able to change colour from the mixer too? Similar thing with mixer channel name- being able to edit it through mixer too.
I understand this PR's scope is limited to addition of the option and linking of colour and names, right?
Also, upon right click, the "Move left/right" seems to be broken. I can still move them with Alt-left/Alt-right, but this behaviour acts strange (exchanging of Mixer channel names?, not entirely clear what's happening). There are a lot of weird issues, this isn;t difficult to reproduce. Also experiment with move left/right after adding a new instrument while auto-track-link is on. This is a **bug**.

Thanks!

spechtstatt commented:
> I noticed a few things, could you clarify these?
> 

Yes, of course - thank you for your thoughts and remarks.

> * Setting an fx mixer colour before linking makes the FX mixer's colour permanent unless the track's colour is changed (not reset, changed). This is a **bug**.

I will look into this - I have added your point to my task list above

> * Could we get a direct "Assign and link to new Mixer channel" option, so we don't have to do it twice? It seems useful. Also an "Link to Mixer channel" in the context menu of track itself?

I think this is a good idea - we should only consider that it may not be used very often because of the automatic adding and assigning of mixer channels - so maybe it is enough to provide only the second option? 

> * After linking, colours are linked but can be edited only from the track's context menu. However, some of us color-code multiple channels consecutively in the mixer itself based on the instrument type, and it could get annoying having to search. Can this be made two-way, being able to change colour from the mixer too? Similar thing with mixer channel name- being able to edit it through mixer too.

I was somehow focused on the idea that the track should control the mixer line (@unfa's original suggestion was to disable the auto link if the mixer name changes)  but because it is in fact a management of a 1:1 relationship there is no real reason to not enable this from both sides - so, yes I think this is also a good idea. 

> * I understand this PR's scope is limited to addition of the option and linking of colour and names, right?

Principally yes - I was just thinking to display some information on the mixer lines because it is collected anyway (as I mentiond above) - like if the line is used at all, how many tracks are linked or if this is linked to a pattern or a song track.

> * Upon right click on a Mixer channel, the "Move left/right" seems to be broken. I can still move them with Alt-left/Alt-right, but this behaviour acts strange (exchanging of Mixer channel names?, not entirely clear what's happening). There are a lot of weird issues, this isn;t difficult to reproduce. Also experiment with move left/right after adding a new instrument while auto-track-link is on. This is a **bug**.

I have the whole sorting issue not really finished yet and just prepared the lock of the move operations and concentrated on the auto link updating. So it is very plausible that there are still flaws but thank you for giving me more detailed feedback about this. 


> 
> Thanks!

spechtstatt commented:
I was thinking that with this functionality enabled it maybe makes sense to always auto link channels by default when assigning them and if they are free.

Because the link would anyway automatically removed if more than one gets assigned. And it could automaticallylink again if there would be only one instrument left.

And it would use the track text/color by default because it is initiated from there.

allejok96 commented:
Tested and worked good.

Makes sense to link by default when an FX is *created* via the menu. Currently LMMS copies the name. That's halfway there :)

Don't do this for assignment to empty FX from the menu, because you might have a template with a "sidechain" FX and you start by assigning one instrument and it does this linking thing and changes the name and color and you be like *hey what's going on how do I rename it back* instead of just ignoring it and continuing to assign like you should.

Likewise I'm skeptical to auto re-link an FX when it becomes only one instrument again, because deleting that instrument may delete the FX without the user realizing it, and effects and stuff may get lost.

Suggestion on toggle button for auto-creating mixer channels for new tracks:
![bild](https://user-images.githubusercontent.com/7693838/170564384-a6f42fdd-a972-4971-976b-d740f6eb1fc9.png)

If there will be more than one option, we could go for a cogwheel button with a menu instead, like on tracks.

Agreeing with Monospace that it would be nice if the ability to set color went both ways. I'd actually like it if you could change the color of an FX channel and it would color multiple tracks in the song editor. But that may be out of context.

It doesn't feel that important to be able to change name of linked channels from the mixer tho.

Marking options grey in the context menu instead of just leaving them out feels better. I know there are lots of things left to be polished. Just mentioning it.

spechtstatt commented:
@allejok96: also thought about multiple tracks linked to a fx line but I found not really a good solution for it regarding the naming.

But maybe it is a solution to link only the color as you suggested.

If there would be a good way to bring both approaches together - of course it could behave different depending on the count of linked tracks (1 = link name & color, n= link only color) but I am not sure if this is a good idea especially if the count changes only temporary during reorganizing.

Regarding multiple options I would probably add a sort button there  because I think it is better to have it not in an automatic fashion. And also an options button where you can set up e.g. if and how the pattern editor should be auto linked and which sorting should be applied when pressing the sort button.

If the sorting is manual I will add the move menu entries again and yes you are absolutely right: it is normally better to grey them out instead of hiding them (don't know why I did it this way 😵)

spechtstatt commented:
@allejok96: also thought about multiple tracks linked to a fx line but I found not really a good solution for it regarding the naming.

But maybe it is a solution to link only the color as you suggested.

If there would be a good way to bring both approaches together - of course it could behave different depending on the count of linked tracks (1 = link name & color, n= link only color) but I am not sure if this is a good idea especially if the count changes only temporary during reorganizing.

Regarding multiple options I would probably add a sort button there  because I think it is better to have it not in an automatic fashion. And also an options button where you can set up e.g. if and how the pattern editor should be auto linked and which sorting should be applied when pressing the sort button.

If the sorting is manual I will add the move menu entries again and yes you are absolutely right: it is normally better to grey them out instead of hiding them (don't know why I did it this way 😵)

spechtstatt commented:
Because I got the impression that there are different needs regarding the auto linking it is now possible to define different settings like: 
- linking one track to one channel vs. linking multiple tracks to one channel
- linking name and color vs linking only the color
- enable the automatic deletion of auto linked channels if there is no track assigned vs. disabling the automatic deletion of auto linked channels
- add channels automatically for the pattern editor like the song editor vs. assigning always the first channel of the pattern editor to all subsequent added pattern tracks (if you want to use only one mixer channel for the pattern editor)
- enable and disable the automatic sorting and also if the song tracks should ordered first or the pattern editor tracks.

I am not sure if this is the right approach or if it would be better to define two or three named sets of settings which correspond to different ways of working - but maybe this is unrealistic. 

Another way would be to reduce the options and split them - e.g. while the sorting currently considers only auto track linked tracks it could be extracted by allowing sorting of all (normal and linked) assigned tracks. (If there are multiple assigned tracks the last  would determine the sort order).

But anyway: with the settings it should now be possible to play around with the various approaches.

spechtstatt commented:
Still some issues with the pattern editor.
